rust/lance-index/src/vector/sq.rs

```diff
@@ -0,0 +1,258 @@
+///
+//
+// TODO: move this to be pub(crate) once we have a better way to test it.
+#[derive(Debug)]
+pub struct ScalarQuantizerImpl<T: ArrowFloatType + Dot + L2> {
+    /// Number of bits for the centroids.
+    ///
+    /// Only support 8, as one of `u8` byte now.
+    pub num_bits: u16,
+
+    /// Distance type.
+    pub metric_type: MetricType,
+
+    pub lower_bound: T::Native,
+    pub upper_bound: T::Native,
+}
+
+impl<T: ArrowFloatType + Dot + L2> ScalarQuantizerImpl<T> {
+    pub fn new(num_bits: u16, metric_type: MetricType) -> Self {
+        Self {
+            num_bits,
+            metric_type,
+            lower_bound: T::MAX,
+            upper_bound: T::MIN,
+        }
+    }
+
+    pub fn with_bounds(
+        num_bits: u16,
+        metric_type: MetricType,
+        bounds: (T::Native, T::Native),
+    ) -> Self {
+        let mut sq = Self::new(num_bits, metric_type);
+        sq.lower_bound = bounds.0;
+        sq.upper_bound = bounds.1;
+        sq
+    }
+
+    pub fn update_bounds(
+        &mut self,
+        vectors: &FixedSizeListArray,
+    ) -> Result<(T::Native, T::Native)> {
+        let data = vectors
+            .values()
+            .as_any()
+            .downcast_ref::<T::ArrayType>()
+            .ok_or(Error::Index {
+                message: format!(
+                    "Expect to be a float vector array, got: {:?}",
+                    vectors.value_type()
+                ),
+                location: location!(),
+            })?
+            .as_slice();
+
+        (self.lower_bound, self.upper_bound) = data
+            .iter()
+            .fold((self.lower_bound, self.upper_bound), |f, v| {
+                (f.0.min(*v), f.1.max(*v))
+            });
+
+        Ok((self.lower_bound, self.upper_bound))
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: ArrowFloatType + Dot + L2 + 'static> ScalarQuantizer for ScalarQuantizerImpl<T> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    async fn transform(&self, data: &dyn Array) -> Result<ArrayRef> {
+        let fsl = data
+            .as_fixed_size_list_opt()
+            .ok_or(Error::Index {
+                message: format!(
+                    "Expect to be a FixedSizeList<float> vector array, got: {:?} array",
+                    data.data_type()
+                ),
+                location: location!(),
+            })?
+            .clone();
+        let data = fsl
+            .values()
+            .as_any()
+            .downcast_ref::<T::ArrayType>()
+            .ok_or(Error::Index {
+                message: format!(
+                    "Expect to be a float vector array, got: {:?}",
+                    fsl.value_type()
+                ),
+                location: location!(),
+            })?
+            .as_slice();
+
+        // TODO: support SQ4
+        let builder: Vec<u8> = data
+            .iter()
+            .map(|v| {
+                let range = self.upper_bound - self.lower_bound;
```

Comment:
Should we have upper/lower bound for each dimension?

Reply:
that will make the contribution of each dimension different